File: Uses-Cases/Tables/get_common.js
// 1. Load your Application Secret and Key from the JSON file or set credentials in another way
// 2. Create an object to connect to the Pdf.Cloud API
// 3. Upload your document file
// 4. Perform the retreiving link annotattions from Pdf document using getPageLinkAnnotations() function
// 5. Check result and perform some actions with result.body object
// 6. Create a new Link Annotation with the required properties
// 7. Append new Link Annotation to the document using postPageLinkAnnotations() function
// 8. Perform some action after successful addition
// All values of variables starting with "YOUR_****" should be replaced by real user values

import credentials from "./../../credentials.json"  with { type: "json" };
import fs from 'node:fs/promises';
import path from 'node:path';
import { PdfApi } from "./../../src/api/api.js";
import { Table } from "../../src/models/table.js";
import { Cell } from "../../src/models/cell.js";
import { FontStyles } from "../../src/models/fontStyles.js";
import { GraphInfo } from "../../src/models/graphInfo.js";
import { Row } from "../../src/models/row.js";
import { TextRect } from "../../src/models/textRect.js";

const configParams = {
    LOCAL_PATH: "C:\\Samples\\",

    PDF_DOCUMENT_NAME: "sample.pdf",

    LOCAL_RESULT_DOCUMENT_NAME: "output_sample.pdf",

    PAGE_NUMBER: 2,     // Your document page number...

    TABLE_ID: "GE5TCOZSGAYCYNRQGUWDINZVFQ3DGMA",
};

const pdfApi = new PdfApi(credentials.id, credentials.key);

const pdfTables = {
    uploadFiles: async function (fileName) {
        const pdfFileData = await fs.readFile(configParams.LOCAL_PATH + fileName);
        await pdfApi.uploadFile(fileName, pdfFileData);
    },

    downloadFiles: async function (local_path, fileName) {
        const changedPdfData = await pdfApi.downloadFile(configParams.PDF_DOCUMENT_NAME);

        const filePath = path.join(local_path, fileName);

        await fs.writeFile(filePath, changedPdfData.body);
        console.log("downloaded: " + filePath);
    },

    uploadDocument: async function () {
        await pdfTables.uploadFiles(configParams.PDF_DOCUMENT_NAME);
    },

    getAllTables: async function () {
        const resultTabs = await pdfApi.getDocumentTables(configParams.PDF_DOCUMENT_NAME);

        if (resultTabs.body.code == 200 && resultTabs.body.tables) {
            if (!Array.isArray(resultTabs.body.tables.list) || resultTabs.body.tables.list.length === 0) {
                throw new Error("Unexpected error : tables is null or empty!!!");
            }
            pdfTables.showTablesInfo(resultTabs.body.tables.list, "all");
            return resultTabs.body.tables.list;
        }
        else
            throw new Error("Unexpected error : can't get tables!!!");
    },

    getTableById: async function () {
        const resultTabs = await pdfApi.getTable(configParams.PDF_DOCUMENT_NAME, configParams.TABLE_ID);

        if (resultTabs.body.code == 200 && resultTabs.body.table) {
            pdfTables.showTablesInfo( [ resultTabs.body.table ], "byId");
            return resultTabs.body.table;
        }
        else
            throw new Error("Unexpected error : can't get tables!!!");
    },

    initTable: function () {
        const numOfCols = 5;
        const numOfRows = 5;

        const headerTextState = {
            font: "Arial Bold",
            fontSize: 11,
            foregroundColor: { a: 0xFF, r: 0xFF, g: 0xFF, b: 0xFF },
            fontStyle: FontStyles.Bold,
            fontFile: null,
            underline: false,
            strikeOut: false,
            superscript: false,
            subscript: false,
        };

        const commonTextState = {
            font: "Arial Bold",
            fontSize: 11,
            foregroundColor: { a: 0xFF, r: 0x70, g: 0x70, b: 0x70 },
            fontFile: null,
            underline: false,
            strikeOut: false,
            superscript: false,
            subscript: false,
        };
    
        const table = new Table();
        table.rows = [];
    
        let colWidths = "";
        for (let c = 0; c < numOfCols; c++)
        {
            colWidths += " 70";
        }
    
        table.columnWidths = colWidths;
    
        const borderTableBorder = new GraphInfo();
        borderTableBorder.color = { a: 0xFF, r: 0x00, g: 0xFF, b: 0x00 };
        borderTableBorder.lineWidth = 0.5;
    
        table.defaultCellBorder = {
            top: borderTableBorder,
            right: borderTableBorder,
            bottom: borderTableBorder,
            left: borderTableBorder,
            roundedBorderRadius: 0
        };
        table.left = 150;
        table.top = 250;
    
        for (let r = 0; r < numOfRows; r++)
        {
            const row = new Row();

            row.cells = [];
    
            for (let c = 0; c < numOfCols; c++)
            {
                const cell = new Cell();
                
                cell.defaultCellTextState = commonTextState;

                if (r == 0)  // header cells
                {
                    cell.backgroundColor = { a: 0xFF, r: 0x80, g: 0x80, b: 0x80 };
                    cell.defaultCellTextState = headerTextState;
                }
                else {
                    cell.backgroundColor = { a: 0xFF, r: 0xFF, g: 0xFF, b: 0xFF };
                };
              
                const textRect = new TextRect();
                if (r == 0)
                    textRect.text = "header #" + c;
                else
                    textRect.text = "value";
                cell.paragraphs = [textRect];

                row.cells.push(cell);
            }
            table.rows.push(row);
        }
        return table;
    },

    addTableOnPage: async function (pageNum) {
        const jsonTable = pdfTables.initTable();

        const resultTabs = await pdfApi.postPageTables(configParams.PDF_DOCUMENT_NAME, pageNum, [ jsonTable ]);

        if (resultTabs.body.code == 200) {
            console.log("Table successfully added!");
            return resultTabs.body.table;
        }
        else
            throw new Error("Unexpected error : can't get tables!!!");
    },

    showTablesInfo: function(tables, prefix) {
        if (Array.isArray(tables) && tables.length > 0)
        {
            tables.forEach(function(table) {
                console.log(prefix +" => id: '" + table.id + "', page: '" + table.pageNum + "', rows: '" + table.rowList.length + "'");
            });
        }
        else
            console.error("showTablesInfo() error: array of tables is empty!")
    },
}

export default pdfTables;

await (async () => {
    await pdfTables.uploadDocument();
    await pdfTables.getAllTables();
    await pdfTables.getTableById();
    await pdfTables.addTableOnPage(configParams.PAGE_NUMBER);
    await pdfTables.downloadFiles( configParams.LOCAL_PATH, configParams.LOCAL_RESULT_DOCUMENT_NAME );
})().catch((error) => { console.log(error.message); });

File: UsesCases/Images/get_common.py
"""
    1. Load your Application Secret and Key from the JSON file or set credentials in another way
    2. Create an object to connect to the Pdf.Cloud API
    3. Initialize parameters (folders and Pdf document) for next actions
    4. Upload your Pdf document file
    5. Perform the retreiving image Id from Pdf document using get_images() function
    6. Perform the reading image from Pdf document using get_image() function and image Id value
    7. Parse response for upload image into temporary folder 
    8. Initialize second Pdf document for inserting uploaded image from first Pdf document
    9. Perform the retreiving image Id from second Pdf document using get_images() function
    10. Initialize parameters for replace image in second Pdf document
    11. Perform replacing image actiont using put_replace_image() function
    12. Check result and perform some actions with response object
    
    All values of variables starting with "YOUR_****" should be replaced by real user values
"""

import os
import shutil
import asposepdfcloud

from urllib.parse import urlparse

pdf_api_client = asposepdfcloud.ApiClient("YOUR_APP_KEY", "YOUR_APP_SECRET")
api = asposepdfcloud.PdfApi(pdf_api_client)

temp_folder = 'TempPdfCloud'
data_path = "C:/Samples/"
output_path = "C:/Samples/Output"

document_name = "Sample-Document-01.pdf"
document_name2 = "Sample-Document-02.pdf"

page_number = 1
def uploadFile(name):
    api.upload_file(temp_folder + '/' + name, data_path + name)

uploadFile(document_name)
uploadFile(document_name2)

opts = {
    "folder" : temp_folder
}

responseImages = api.get_images(document_name, page_number, **opts)

image_id = responseImages.images.list[0].id

responseImage = api.get_image(document_name, image_id)

print (responseImage)

file_name_link = responseImage.image.links[0].href

a = urlparse(file_name_link)
image_file_name = os.path.basename(a.path)
image_file_path = temp_folder +"/" + image_file_name

resImage = api.download_file(file_name_link)
shutil.move(resImage, image_file_path)

uploadFile(api, image_file_path, image_file_name)

opts = {
    "image_file_path" : self.temp_folder + '/' + image_file_name,
              "folder" : self.temp_folder
        }
        response = self.pdf_api.put_replace_image(file_name, image_id, **opts)

response = api.put_replace_image(file_name2, image_id2, **opts)

print (response)